File: kranji-studio/src/main/resources/homing/js/kranji/studio/articles/ArticleShelfSecretaryModule.js
// =============================================================================
// ArticleShelfSecretaryModule — the Secretary for the article workspace's shelf.
//
// Three panes, one conversation. Roots picks a folder; Navigator picks a draft
// inside it; Drafts shows what was picked. PickRoot and PickDraft go in,
// ShelfChanged comes back out to everybody.
//
// ONE bus, not two. Two parties would mean two declarations and two joins per
// widget, and a widget that forgot one would go quiet with nothing to show for
// it. `what` says which half moved, so a pane reacts only to its own.
//
// The whole state rides on every broadcast: { what, root, draft, name }. A pane
// opened after a root was chosen is correct on the next change rather than
// needing a resync handshake - the same reason the gloss bus carries its whole
// selection.
//
// Pure (state, envelope) -> { newState, actions }. No DOM, no captures.
// =============================================================================

var ArticleShelfSecretary = {

    initial: {
        // '' is "nothing chosen yet", which a pane shows as its own default -
        // the first root on the shelf - rather than as an empty screen.
        root: '',
        draft: '',
        name: '',
        recentUnknown: []
    },

    behavior: function (state, envelope) {
        var msg = envelope.message;

        // Choosing a root CLEARS the draft. Draft ids are relative to their
        // root, so an id from one folder means nothing in another - and the
        // bad case is not that it resolves to nothing, it is that it resolves
        // to a different file sitting at the same relative path.
        if (msg.kind === "PickRoot") {
            var next = { root: msg.root || '', draft: '', name: '',
                         recentUnknown: state.recentUnknown };
            return {
                newState: next,
                actions: [{
                    kind: "BroadcastToMembers",
                    message: {
                        kind: "ShelfChanged", what: "root",
                        root: next.root, draft: '', name: ''
                    }
                }]
            };
        }

        if (msg.kind === "PickDraft") {
            // The root travels with the draft rather than being read from
            // state, because the pane that picked it knows which tree it was
            // looking at and this actor does not.
            var withDraft = {
                root: msg.root || state.root,
                draft: msg.draft || '',
                name: msg.name || '',
                recentUnknown: state.recentUnknown
            };
            return {
                newState: withDraft,
                actions: [{
                    kind: "BroadcastToMembers",
                    message: {
                        kind: "ShelfChanged", what: "draft",
                        root: withDraft.root,
                        draft: withDraft.draft,
                        name: withDraft.name
                    }
                }]
            };
        }

        // For a pane that arrived late and would rather ask than wait for the
        // next change. Nothing moves; the whole picture is re-broadcast.
        if (msg.kind === "WhatIsChosen") {
            return {
                newState: state,
                actions: [{
                    kind: "BroadcastToMembers",
                    message: {
                        kind: "ShelfChanged", what: "resync",
                        root: state.root, draft: state.draft, name: state.name
                    }
                }]
            };
        }

        // Unknown kinds are remembered and otherwise ignored. A bus that threw
        // would couple every producer to every consumer's vocabulary.
        var seen = state.recentUnknown.concat([
            { kind: msg.kind, from: envelope.from }
        ]);
        return {
            newState: {
                root: state.root,
                draft: state.draft,
                name: state.name,
                recentUnknown: seen.slice(-10)
            },
            actions: []
        };
    }
};

File: kranji-studio/src/main/java/kranji/studio/articles/ArticleShelfSecretaryModule.java
package kranji.studio.articles;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The Secretary for the article workspace's shelf.
 *
 * <p>Three panes and one conversation: a root is chosen, and then a draft
 * inside it. {@code PickRoot} and {@code PickDraft} go in;
 * {@code ShelfChanged} comes back out to everybody.</p>
 *
 * <p>One bus rather than two, for the reason the gloss workbench gives next
 * door: two parties would mean two declarations and two joins per widget, and
 * a widget that forgot one would go quiet with nothing to show for it. What
 * changed is a field on the message, so a pane reacts to the half it cares
 * about.</p>
 *
 * <p>Every broadcast carries the whole state. A pane opened after a root was
 * chosen is then correct on the next change rather than needing a resync
 * handshake — the same reason the gloss bus carries its whole selection.</p>
 *
 * <p><b>Choosing a root clears the draft.</b> Draft ids are relative to their
 * root, so an id from one folder means nothing in another and would resolve to
 * either nothing or, far worse, a different file that happens to sit at the
 * same relative path.</p>
 */
public record ArticleShelfSecretaryModule()
        implements DomModule<ArticleShelfSecretaryModule> {

    /** The pure (state, envelope) → step function. */
    public record ArticleShelfSecretary()
            implements Exportable._Constant<ArticleShelfSecretaryModule> {}

    public static final ArticleShelfSecretaryModule INSTANCE =
            new ArticleShelfSecretaryModule();

    @Override
    public ImportsFor<ArticleShelfSecretaryModule> imports() {
        return ImportsFor.<ArticleShelfSecretaryModule>builder().build();
    }

    @Override
    public ExportsOf<ArticleShelfSecretaryModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new ArticleShelfSecretary()));
    }
}
